pkg/resource/reference_resolver.go

```diff
@@ -0,0 +1,282 @@
+}
+
+// AttributeReferencer is an interface for referencing and resolving
+// cross-resource attribute references. See
+// https://github.com/crossplaneio/crossplane/blob/master/design/one-pager-cross-resource-referencing.md
+// for more information
+type AttributeReferencer interface {
+
+	// GetStatus looks up the referenced objects in K8S api and returns a list
+	// of ReferenceStatus
+	GetStatus(context.Context, CanReference, client.Reader) ([]ReferenceStatus, error)
+
+	// Build retrieves referenced resource, as well as other non-managed
+	// resources (like a `Provider`), and builds the referenced attribute
+	Build(context.Context, CanReference, client.Reader) (string, error)
+
+	// Assign accepts a managed resource object, and assigns the given value to the
+	// corresponding property
+	Assign(CanReference, string) error
+}
+
+// A ManagedReferenceResolver resolves the references to other managed
+// resources, by looking them up in the Kubernetes API server. The references
+// are the fields in the managed resource that implement AttributeReferencer
+// interface and have
+// `attributeReferencerTagName:"managedResourceStructTagPackageName"` tag
+type ManagedReferenceResolver interface {
+	ResolveReferences(context.Context, CanReference) error
+}
+
+// APIManagedReferenceResolver resolves implements ManagedReferenceResolver interface
+type APIManagedReferenceResolver struct {
+	client client.Client
+}
+
+// NewReferenceResolver returns a new APIManagedReferenceResolver
+func NewReferenceResolver(c client.Client) *APIManagedReferenceResolver {
+	return &APIManagedReferenceResolver{c}
+}
+
+// ResolveReferences resolves references made to other managed resources
+func (r *APIManagedReferenceResolver) ResolveReferences(ctx context.Context, res CanReference) (err error) {
+	// retrieve all the referencer fields from the managed resource
+	referencers, err := findAttributeReferencerFields(res, false)
+	if err != nil {
+		// if there is an error it should immediately panic, since this means an
+		// attribute is tagged but doesn't implement AttributeReferencer
+		panic(err)
+	}
+
+	// if there are no referencers exit early
+	if len(referencers) == 0 {
+		return nil
+	}
+
+	// this recovers from potential panics during execution of
+	// AttributeReferencer methods
+	defer func() {
+		if r := recover(); r != nil {
+			err = errors.Errorf(errPanicedResolving, r)
+		}
+	}()
+
+	// make sure that all the references are ready
+	allStatuses := []ReferenceStatus{}
+	for _, referencer := range referencers {
+		statuses, err := referencer.GetStatus(ctx, res, r.client)
+		if err != nil {
+			return err
+		}
+
+		allStatuses = append(allStatuses, statuses...)
+	}
+
+	if err := newReferenceAccessErr(allStatuses); err != nil {
+		return err
+	}
+
+	// build and assign the attributes
+	for _, referencer := range referencers {
+		val, err := referencer.Build(ctx, res, r.client)
+		if err != nil {
+			return errors.WithMessage(err, errBuildAttribute)
```

Comment:
https://godoc.org/github.com/pkg/errors#Wrap

Nitpick - the prevalent convention is to use `Wrap` rather than `WithMessage`. It seems they're identical except that `Wrap` includes a stack trace and `WithMessage` does not.